odoo/modules/registry.py

```diff
@@ -325,71 +324,134 @@ def field_computed(self):
+
+        def transitive_triggers(field, prefix=(), seen=()):
+            if field in seen or field not in triggers:
                 return
-            for seq1 in dependencies.get(field, ()):
-                yield seq1
-                for seq2 in transitive_dependencies(seq1[-1], seen + [field]):
-                    yield concat(seq1[:-1], seq2)
+            for path, targets in triggers[field].items():
+                full_path = concat(prefix, path)
+                yield full_path, targets
+                for target in targets:
+                    yield from transitive_triggers(target, full_path, seen + (field,))
 
         def concat(seq1, seq2):
             if seq1 and seq2:
                 f1, f2 = seq1[-1], seq2[0]
-                if f1.type == 'one2many' and f2.type == 'many2one' and \
-                        f1.model_name == f2.comodel_name and f1.inverse_name == f2.name:
+                if (
+                    f1.type == 'many2one' and f2.type == 'one2many'
+                    and f1.name == f2.inverse_name
+                    and f1.model_name == f2.comodel_name
+                    and f1.comodel_name == f2.model_name
+                ):
                     return concat(seq1[:-1], seq2[1:])
             return seq1 + seq2
 
-        # determine triggers based on transitive dependencies
-        triggers = {}
-        for field in dependencies:
-            for path in transitive_dependencies(field):
-                if path:
-                    tree = triggers
-                    for label in reversed(path):
-                        tree = tree.setdefault(label, {})
-                    tree.setdefault(None, OrderedSet()).add(field)
+        tree = TriggerTree()
+        for path, targets in transitive_triggers(field):
+            current = tree
+            for label in path:
+                current = current.increase(label)
+            if current.root:
+                current.root.update(targets)
+            else:
+                current.root = OrderedSet(targets)
 
-        return triggers
+        self._field_trigger_trees[field] = tree
+
+        return tree
 
     @lazy_property
-    def fields_modifying_relations(self):
-        '''
-        Return the union of the set of relational fields that are dependencies
-        of other fields, with the set of non-relational fields that are
-        dependencies of relational fields.
-        '''
-        result = set()
-
-        for field in self.field_triggers:
-            # If the field is itself a relational field, it is also considered
-            # as triggering relational fields.
-            if field.relational or self.field_inverses[field]:
-                result.add(field)
+    def _field_trigger_trees(self):
+        # internal cache of method get_field_trigger_tree()
+        return {}
+
+    @lazy_property
+    def _field_triggers(self):
+        # {field: {path: {field}}}
+        triggers = defaultdict(lambda: defaultdict(OrderedSet))
```

Comment:
I think that we could actually use a `list` here, since duplication is dealt with in the trigger trees.  And there would be duplicates only if dependencies themselves have duplicates.